internal/telemetry/telemetry.go

```diff
@@ -169,42 +230,80 @@ func newState() *statev1.TelemetryState {
 	}
 }
 
-func (r *reporter) shouldReport(state *statev1.TelemetryState) bool {
-	for _, v := range []string{noTelemetryEnvVar, doNotTrackEnvVar} {
-		if disabledByEnvVar(v) {
-			r.logger.Info("Telemetry disabled")
-			return false
+func (r *segmentReporter) start() {
+	defer func() {
+		// don't let a panic in this goroutine crash the whole app.
+		if err := recover(); err != nil {
+			r.logger.Debug("Telemetry panic", zap.Any("cause", err))
 		}
-	}
+	}()
 
-	r.logger.Info(fmt.Sprintf("Anonymous telemetry enabled. Disable via the config file or by setting the %s=1 environment variable", noTelemetryEnvVar))
+	r.reportServerLaunch()
 
-	if state == nil || state.LastTimestamp == nil {
-		return true
+	for event := range r.eventChan {
+		switch t := event.Data.(type) {
+		case *telemetryv1.Event_ApiActivity_:
+			r.reportAPIActivity(t.ApiActivity)
+		default:
+			r.logger.Debug(fmt.Sprintf("Unhandled telemetry event type %T", t))
+		}
 	}
+}
 
-	lastTS := state.LastTimestamp.AsTime()
-	return time.Since(lastTS) > minReportInterval
+func (r *segmentReporter) Intercept() Interceptors {
+	return newStatsInterceptors(r, r.reportInterval, r.shutdownChan)
 }
 
-func disabledByEnvVar(name string) bool {
-	v, ok := os.LookupEnv(name)
-	// if the var is not defined, assume consent.
-	if !ok {
-		return false
+func (r *segmentReporter) reportServerLaunch() {
+	event := buildServerLaunch(r.store)
+	if err := r.send("server_launch", event); err != nil {
+		r.logger.Debug("Failed to send server launch event", zap.Error(err))
 	}
+}
 
-	set, err := strconv.ParseBool(v)
-	if err != nil {
-		// err on the side of caution and assume no consent.
+func (r *segmentReporter) Report(event *telemetryv1.Event) bool {
+	select {
+	case r.eventChan <- event:
 		return true
+	default:
+		return false
 	}
+}
 
-	return set
+func (r *segmentReporter) reportAPIActivity(event *telemetryv1.Event_ApiActivity) {
+	if err := r.send("api_activity", event); err != nil {
+		r.logger.Debug("Failed to send API activity event", zap.Error(err))
+	}
 }
 
-func (r *reporter) writeState(state *statev1.TelemetryState) error {
-	stateBytes, err := protojson.Marshal(state)
+func (r *segmentReporter) Stop() error {
+	var err error
+	r.closeOnce.Do(func() {
+		close(r.shutdownChan)
+		close(r.eventChan)
+		r.reportServerStop()
```

Comment:
Perhaps we could report the API activity here too? Otherwise if we stop the server after 59 minutes we lose a bunch of stats.

We might be able to move the request count onto the API activity event in that case.

Reply:
The tricky part here is that interceptor runs in a different goroutine. When the system is shutting down, it's difficult to reliably  synchronize the reporter and the interceptor without resorting to some convoluted blocking code (which could hold up the process from shutting down properly). Because the API and UA stats are complex, they can't be shared between goroutines without locking -- which is not ideal for performance. That is why, as a compromise, the total request count is declared as a global variable. That way if the PDP is short-lived, we still get some idea of usage.